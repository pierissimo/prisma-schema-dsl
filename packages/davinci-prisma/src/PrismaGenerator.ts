import {
  createDataSource,
  createModel,
  createObjectField,
  createScalarField,
  createSchema,
  createView,
  DataSource,
  Enum,
  Generator,
  Model,
  ObjectField,
  print,
  ScalarField,
  ScalarType,
  Schema,
  View,
} from '@pmaltese/prisma-schema-generator'
import { ClassType, DecoratorId, reflect, TypeValue } from '@davinci/reflector'
import {
  FieldDecoratorMeta,
  ModelDecoratorMeta,
  ObjectDecoratorOptions,
  ScalarDecoratorOptions,
} from './types'

const scalarTypes = Object.values(ScalarType)

export class PrismaGenerator {
  datasource: DataSource
  schema: Schema
  generators: Array<Generator>
  enums: Array<Enum>
  classToModelMap = new Map<ClassType, Model>()
  classToViewMap = new Map<ClassType, View>()

  public reflectModels(classTypes: Array<ClassType>) {
    classTypes.forEach((classType) => {
      this.reflectModel(classType)
    })

    return this
  }

  public reflectModel(classType: ClassType) {
    const classReflection = reflect(classType)
    const modelDecorator: ModelDecoratorMeta = classReflection.decorators.find(
      (d) => d[DecoratorId] === 'davinci.prisma.model' || d[DecoratorId] === 'davinci.prisma.view',
    )
    const isModel = modelDecorator[DecoratorId] === 'davinci.prisma.model'

    const fields = classReflection.properties.reduce(
      (acc: Array<ScalarField | ObjectField>, propReflection) => {
        const fieldDecorator: FieldDecoratorMeta = propReflection.decorators.find(
          (d) => d[DecoratorId] === 'davinci.prisma.field',
        )

        if (fieldDecorator) {
          const fieldName = fieldDecorator.data?.name ?? propReflection.name
          /* if (typeof fieldDecorator.data?.type === 'string') {
            throw new Error('Syntax not yet supported')
          } */

          const isArray = Array.isArray(fieldDecorator.data?.type)
          const prismaTypeDefinition = this.parsePrismaType(
            fieldDecorator.data?.type ?? propReflection.type,
          )

          let field: ScalarField | ObjectField

          if (prismaTypeDefinition.isScalar) {
            const fieldDecoratorData = fieldDecorator.data as ScalarDecoratorOptions
            field = createScalarField({
              name: fieldName,
              type: prismaTypeDefinition.type as ScalarType,
              nativeMapping: fieldDecoratorData?.nativeMapping,
              isList: fieldDecoratorData?.isList ?? isArray,
              isRequired: fieldDecoratorData?.required,
              isUnique: fieldDecoratorData?.isUnique,
              isId: fieldDecoratorData?.isId,
              isUpdatedAt: fieldDecoratorData?.isUpdatedAt,
              defaultValue: fieldDecoratorData?.defaultValue,
              documentation: fieldDecoratorData?.documentation,
              isForeignKey: fieldDecoratorData?.isForeignKey,
            })
          } else {
            const fieldDecoratorData = fieldDecorator.data as ObjectDecoratorOptions
            field = createObjectField({
              name: fieldName,
              type: prismaTypeDefinition.type as ScalarType,
              nativeMapping: fieldDecoratorData?.nativeMapping,
              isList: fieldDecoratorData?.isList ?? isArray,
              isRequired: fieldDecoratorData?.required,
              relationName: fieldDecoratorData?.relationName,
              relationFields: fieldDecoratorData?.relationFields,
              relationReferences: fieldDecoratorData?.relationReferences,
              relationOnDelete: fieldDecoratorData?.relationOnDelete,
              relationOnUpdate: fieldDecoratorData?.relationOnUpdate,
              documentation: fieldDecoratorData?.documentation,
            })
          }

          acc.push(field)
        }

        return acc
      },
      [],
    )

    const entity = isModel
      ? createModel({
          name: modelDecorator.data?.name ?? classReflection.name,
          fields,
          map: modelDecorator.data?.map,
          indexes: modelDecorator.data?.indexes,
          uniqueIndexes: modelDecorator.data?.uniqueIndexes,
          fullTextIndexes: modelDecorator.data?.fullTextIndexes,
        })
      : createView({
          name: modelDecorator.data?.name ?? classReflection.name,
          fields,
          map: modelDecorator.data?.map,
          uniqueIndexes: modelDecorator.data?.uniqueIndexes,
        })
    if (isModel) {
      this.classToModelMap.set(classType, entity)
    } else {
      this.classToViewMap.set(classType, entity)
    }

    return {
      type: isModel ? 'model' : 'view',
      entity,
    }
  }

  public setDatasource({ name, provider, url, shadowDatabaseUrl, relationMode }: DataSource) {
    this.datasource = createDataSource({
      name,
      provider,
      url,
      shadowDatabaseUrl,
      relationMode,
    })

    return this
  }

  public setGenerators(generators: Array<Generator>) {
    this.generators = generators

    return this
  }

  public setEnums(enums: Array<Enum>) {
    this.enums = enums

    return this
  }

  public createSchema() {
    this.schema = createSchema({
      models: Array.from(this.classToModelMap.values()),
      enums: this.enums ?? [],
      dataSource: this.datasource,
      generators: this.generators,
      views: Array.from(this.classToViewMap.values()),
    })

    return this
  }

  public async generate() {
    if (!this.schema) {
      this.createSchema()
    }
    const prismaString = await print(this.schema)

    return { schema: this.schema, prismaString }
  }

  private parsePrismaType(type: TypeValue | ScalarType) {
    let isScalar = true
    let parsedType
    if (scalarTypes.includes(type as ScalarType)) {
      parsedType = type as ScalarType
    } else if (type === Date) {
      parsedType = ScalarType.DateTime
    } else if (type === String) {
      parsedType = ScalarType.String
    } else if (type === Number) {
      parsedType = ScalarType.Float
    } else if (type === Boolean) {
      parsedType = ScalarType.Boolean
    } else if (typeof type === 'function') {
      const { entity } = this.reflectModel(type as ClassType)
      parsedType = entity.name
      isScalar = false
    } else if (typeof type === 'string') {
      parsedType = type
      isScalar = false
    } else {
      throw new Error('Unrecognized type')
    }

    return { type: parsedType, isScalar }
  }
}
