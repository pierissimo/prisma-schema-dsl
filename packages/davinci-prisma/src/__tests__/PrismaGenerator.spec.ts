import { prisma } from '../index'
import { PrismaGenerator } from '../PrismaGenerator'
import { DataSourceProvider, Enum, Generator } from '@pmaltese/prisma-schema-generator'

describe('PrismaGenerator', () => {
  test('Models - should correctly generate the schema for primitive field types', async () => {
    @prisma.model()
    class Customer {
      @prisma.field({ required: true, isId: true })
      id: string

      @prisma.field({ required: true })
      firstname: string

      @prisma.field({ required: true })
      lastname: string

      @prisma.field()
      age: number

      @prisma.field()
      isAdult: boolean

      @prisma.field()
      createdAt: Date

      @prisma.field({ isUpdatedAt: true })
      updatedAt: Date
    }
    const prismaGenerator = new PrismaGenerator()

    const { schema } = await prismaGenerator.reflectModels([Customer]).generate()
    expect(schema).toMatchObject({
      models: [
        {
          name: 'Customer',
          fields: [
            {
              name: 'id',
              isList: false,
              isRequired: true,
              isUnique: false,
              kind: 'scalar',
              type: 'String',
              isId: true,
              isUpdatedAt: false,
              default: null,
              isForeignKey: false,
            },
            {
              name: 'firstname',
              isList: false,
              isRequired: true,
              isUnique: false,
              kind: 'scalar',
              type: 'String',
              isId: false,
              isUpdatedAt: false,
              default: null,
              isForeignKey: false,
            },
            {
              name: 'lastname',
              isList: false,
              isRequired: true,
              isUnique: false,
              kind: 'scalar',
              type: 'String',
              isId: false,
              isUpdatedAt: false,
              default: null,
              isForeignKey: false,
            },
            {
              name: 'age',
              isList: false,
              isRequired: false,
              isUnique: false,
              kind: 'scalar',
              type: 'Float',
              isId: false,
              isUpdatedAt: false,
              default: null,
              isForeignKey: false,
            },
            {
              name: 'isAdult',
              isList: false,
              isRequired: false,
              isUnique: false,
              kind: 'scalar',
              type: 'Boolean',
              isId: false,
              isUpdatedAt: false,
              default: null,
              isForeignKey: false,
            },
            {
              name: 'createdAt',
              isList: false,
              isRequired: false,
              isUnique: false,
              kind: 'scalar',
              type: 'DateTime',
              isId: false,
              isUpdatedAt: false,
              default: null,
              isForeignKey: false,
            },
            {
              name: 'updatedAt',
              isList: false,
              isRequired: false,
              isUnique: false,
              kind: 'scalar',
              type: 'DateTime',
              isId: false,
              isUpdatedAt: true,
              default: null,
              isForeignKey: false,
            },
          ],
        },
      ],
      views: [],
    })
  })

  test('Models - should correctly generate the relations and indexes', async () => {
    @prisma.model({
      indexes: [{ fields: [{ name: 'id', sort: 'asc' }] }],
      uniqueIndexes: [{ fields: [{ name: 'id' }] }],
      fullTextIndexes: [{ fields: [{ name: 'firstname' }, { name: 'lastname' }] }],
    })
    class Customer {
      @prisma.field({ required: true, isId: true })
      id: string

      @prisma.field({ required: true })
      firstname: string

      @prisma.field({ required: true })
      lastname: string

      @prisma.field({
        type: 'CustomerToken',
        isList: true,
        required: true,
      })
      tokens: Array<CustomerToken>
    }

    @prisma.model()
    class CustomerToken {
      @prisma.field()
      customerId: string

      @prisma.field({
        type: 'Customer',
        relationFields: ['customerId'],
        relationReferences: ['id'],
      })
      customer: Customer
    }
    const prismaGenerator = new PrismaGenerator()

    const { schema } = await prismaGenerator.reflectModels([Customer, CustomerToken]).generate()
    expect(schema).toMatchObject({
      models: [
        {
          name: 'Customer',
          fields: [
            {
              name: 'id',
              // [...]
            },
            {
              name: 'firstname',
              // [...]
            },
            {
              name: 'lastname',
              // [...]
            },
            {
              name: 'tokens',
              isList: true,
              isRequired: true,
              kind: 'object',
              type: 'CustomerToken',
              relationName: null,
              relationToFields: [],
              relationToReferences: [],
              relationOnDelete: 'NONE',
              relationOnUpdate: 'NONE',
            },
          ],
          indexes: [
            {
              fields: [
                {
                  name: 'id',
                  sort: 'asc',
                },
              ],
            },
          ],
          fullTextIndexes: [
            {
              fields: [
                {
                  name: 'firstname',
                },
                {
                  name: 'lastname',
                },
              ],
            },
          ],
          uniqueIndexes: [
            {
              fields: [
                {
                  name: 'id',
                },
              ],
            },
          ],
        },
        {
          name: 'CustomerToken',
          fields: [
            {
              name: 'customerId',
              // [...]
            },
            {
              name: 'customer',
              isList: false,
              isRequired: false,
              kind: 'object',
              type: 'Customer',
              relationName: null,
              relationToFields: ['customerId'],
              relationToReferences: ['id'],
              relationOnDelete: 'NONE',
              relationOnUpdate: 'NONE',
            },
          ],
        },
      ],
    })
  })

  test('DataSource - should correctly set up the data source', async () => {
    const prismaGenerator = new PrismaGenerator()
    const { schema } = await prismaGenerator
      .setDatasource({
        name: 'myDatasource',
        environmentVariable: 'DB_URL',
        provider: DataSourceProvider.PostgreSQL,
      })
      .generate()

    expect(schema).toMatchObject({
      dataSource: {
        name: 'myDatasource',
        provider: 'postgresql',
        url: {
          name: 'DB_URL',
        },
      },
    })
  })

  test('Generator - should correctly set up the generators', async () => {
    const prismaGenerator = new PrismaGenerator()
    const generatorsConfig: Array<Generator> = [
      {
        name: 'client1',
        provider: 'prisma-client-js',
        // output: './src/generated/prisma-client-js',
        previewFeatures: ['fullTextSearch', 'fullTextIndex', 'clientExtensions'],
      },
      {
        name: 'client2',
        provider: 'prisma-client-js',
        output: './src/generated/prisma-client-js',
        previewFeatures: ['fullTextSearch', 'fullTextIndex', 'clientExtensions'],
        binaryTargets: ['linux-musl'],
      },
    ]
    const { schema } = await prismaGenerator.setGenerators(generatorsConfig).generate()

    expect(schema).toMatchObject({
      generators: generatorsConfig,
    })
  })

  test('Enums - should correctly set up the enums', async () => {
    const prismaGenerator = new PrismaGenerator()
    const enumsConfig: Array<Enum> = [{ name: 'myEnum', values: ['ONE', 'TWO'] }]
    const { schema } = await prismaGenerator.setEnums(enumsConfig).generate()

    expect(schema).toMatchObject({
      enums: enumsConfig,
    })
  })

  test('Schema - should correctly generate a prisma string', async () => {
    @prisma.model()
    class Customer {
      @prisma.field({ required: true, isId: true })
      id: string

      @prisma.field({ required: true })
      firstname: string

      @prisma.field({ required: true })
      lastname: string

      @prisma.field({
        type: 'CustomerToken',
        isList: true,
        required: true,
      })
      tokens: Array<CustomerToken>
    }

    @prisma.model()
    class CustomerToken {
      @prisma.field()
      customerId: string

      @prisma.field({
        type: 'Customer',
        relationFields: ['customerId'],
        relationReferences: ['id'],
      })
      customer: Customer
    }

    const prismaGenerator = new PrismaGenerator()
      .setGenerators([
        {
          name: 'client1',
          provider: 'prisma-client-js',
          previewFeatures: ['fullTextSearch', 'fullTextIndex', 'clientExtensions'],
        },
      ])
      .setDatasource({
        name: 'myDatasource',
        environmentVariable: 'DB_URL',
        provider: DataSourceProvider.PostgreSQL,
      })
      .reflectModels([Customer, CustomerToken])

    const { prismaString } = await prismaGenerator
      .reflectModels([Customer, CustomerToken])
      .generate()

    expect(prismaString).toMatchSnapshot()
  })
})
