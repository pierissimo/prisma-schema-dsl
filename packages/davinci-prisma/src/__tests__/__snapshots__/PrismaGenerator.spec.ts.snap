// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrismaGenerator Schema - should correctly generate a prisma string 1`] = `
"datasource myDatasource {
  provider = \\"postgresql\\"
  url      = env(\\"DB_URL\\")
}

generator client1 {
  provider        = \\"prisma-client-js\\"
  previewFeatures = [\\"fullTextSearch\\", \\"fullTextIndex\\", \\"clientExtensions\\"]
}

model Customer {
  id             String          @id
  firstname      String
  lastname       String
  storyOfHisLife String?         @myDatasource.VarChar(10000)
  tokens         CustomerToken[]
}

model CustomerToken {
  customerId String?
  customer   Customer? @relation(fields: [customerId], references: [id])
}
"
`;
